### ISSUE TITLE
runtime: determine global services and boostrapping side-effects

### ISSUE THREAD

**Issue by x1B** [LaxarJS/laxar#310]
After #303, a certain specific set of services is exposed as global to the `whenServicesReady` callback:

```
configuration
globalEventBus
i18n
log
storage
tooling
```

Should we add/remove services?

Also, the `laxar.bootstrap` call contains two setup-calls that we might want to move out to the application, since they are not relevant to all users:

```
globalEventBus.setErrorHandler( createLogErrorHandler( log ) );
ensureInstanceId( log, storage );
```

If these are universally applicable, they should be moved into the service factories. If not, they should go to user-land. With the error handler, we might alternatively decide to just make it the new default error handler (what would be the drawbacks?).

**Comment by alex3683** [comment]
Service | available for adapter | available as widget injection | decorated for widget instance
-------- | -------------------- | --------- | ------------------ | ----------------------------------
context | :x: | :heavy_check_mark: (axContext) | :heavy_check_mark: (axContext)
configuration | :heavy_check_mark: | :heavy_check_mark: (axConfiguration) | :x:
browser | :x: | :x: | :x:
toolingCollectors | :x: | :x: | :x:
log | :heavy_check_mark: | :heavy_check_mark: (axGlobalLog) | :heavy_check_mark: (axLog)
storage | :heavy_check_mark: | :heavy_check_mark: (axGlobalStorage) | :heavy_check_mark: (axStorage)
i18n | :heavy_check_mark: | :heavy_check_mark: (axI18n) | :x:
timer | :x: | :x: | :x:
paths | :x: | :x: | :x:
heartbeat | :heavy_check_mark: | :heavy_check_mark: (axHeartbeat) | :x:
eventBus | :heavy_check_mark: | :heavy_check_mark: (axGlobalEventBus) | :heavy_check_mark: (axEventBus)
themeManager | :x: | :x: | :x:
cssLoader | :x: | :x: | :x:
layoutLoader | :x: | :x: | :x:
pageLoader | :x: | :x: | :x:
localeManager | :x: | :x: | :x:
controlsService | :x: | :x: | :x:
visibilityManager | :x: | :x: | :x:
widgetLoader | :x: | :x: | :x:
flowService | :x: | :heavy_check_mark: (axFlowService) | :x:
pageService | :heavy_check_mark: | :x: | :x:
toolingProviders | :heavy_check_mark: | :heavy_check_mark: (axToolingProviders) | :x:

**Comment by x1B** [comment]
It should also be possible to add configure a laxarjs application with `tooling.enabled` (boolean). If `true`, this instance services (those that are also available to the adapters) will be made accessible at `[global].laxarInstances[name]` where `[global]` is e.g. `window` and `name` is the configuration value for `name`.

**Comment by alex3683** [comment]
#306 will be fixed with this issue

**Comment by alex3683** [comment]
Implemented on `master` (`v2.0.0`)

### Breaking Change

The set of services made available to widget adapters and widgets was redefined and accordingly implemented. The complete list can be seen above. Additionally widget adapters are free to add their own custom services e.g. when necessary for support of their respective technology.

Additionally the `whenServicesReady` callback for `laxar.bootstrap` was removed again. We found that it would be too easy to misuse that api for application implementation and work against the widgets and activities architecture. Since we nevertheless need some kind of public access to laxar apis for tooling, the configuration option explained by @x1B was implemented. When tooling support is needed simply set the configuration of `tooling.enabled` to `true` and access the apis under the global object (i.e. `window`) as `global.laxarInstances[name]` where name is the name of the application as it was configured.
